lib/Fencer/Rules.hs

```diff
@@ -100,6 +118,26 @@ loadRulesFromDirectory
         dirs <- filterM isDirectory other
         (files ++) <$> concatMapM listAllFiles dirs
 
+-- | Perform validation checks to make sure the behavior matches that
+-- of @lyft/ratelimit@.
+validatePotentialDomains
+  :: [Either LoadRulesError (Maybe DomainDefinition)]
+  -> Either [LoadRulesError] [DomainDefinition]
+validatePotentialDomains res = case partitionEithers res of
+  (errs@(_:_), _    ) -> Left errs
+  ([]        , mRules) -> do
+    -- check if there are any duplicate domains
+    let
+      rules = catMaybes mRules
+      groupedRules :: [NonEmpty DomainDefinition] = NE.groupBy
+        ((==) `on` (unDomainId . domainDefinitionId))
+        (NE.fromList $ sortOn domainDefinitionId rules)
```

Comment:
Does performance matter? If so, then it's more efficient to just manually insert rules into a `Map` (or if those ids are essentially integers, then in `IntMap`) after checking that a rule is not already in the map (in which case return `Left` immediately without doing anything with the remaining rules).

But if lists of rules are not long or this part is never going to be a bottleneck, then the current definition is fine.

Reply:
I'd argue this will never be a bottleneck. Even if there will be millions of domain definitions, which I doubt, this should be really fast. The complexity of sorting here is O(n log n). This is ran just once on the application start up.

Reply:
Ok, then ignore the suggestion.